### ISSUE TITLE
DiffusionModel.compute_metrics: per-sample SNR weights broadcast against the whole batch (loss = mean(w) * mean(err))

### ISSUE THREAD

**Issue by LarsKue** [bayesflow-org/bayesflow#787]
## Summary

`DiffusionModel.compute_metrics` pairs every sample's noise-level weight with the error of **every** sample in the batch, not with its own. The loss that is optimized is therefore `mean(weights) * mean(error)` instead of `mean(weights * error)`. This happens for any input shape, vectors included.

Found in bayesflow 2.0.13; the same code is on `main` today (`bayesflow/networks/inference/diffusion/diffusion_model.py`, lines ~184-261).

## Cause

```python
log_snr_t = expand_right_as(log_snr_t, x)                    # (B, 1)  or (B, 1, 1, 1)
weights_for_snr = self.noise_schedule.get_weights_for_snr(log_snr_t=log_snr_t)   # same shape
...
loss = weights_for_snr * ops.mean(loss_mask * (noise_pred - eps_t) ** 2, axis=-1)  # (B,) or (B, H, W)
loss = weighted_mean(loss, sample_weight)
```

The error is reduced over the last axis only, while the weights keep a trailing singleton for every non-batch axis. Right-aligned broadcasting then gives:

- vectors: `(B, 1) * (B,) -> (B, B)`
- images: `(B, 1, 1, 1) * (B, H, W) -> (B, B, H, W)`

`weighted_mean` averages everything, so the result factorizes into `mean(w) * mean(err)`. The same pattern appears in the `"velocity"` and `"F"` branches.

## Minimal reproduction

```python
import numpy as np
import keras
from bayesflow.utils import expand_right_as

for shape in [(4, 6), (4, 8, 8, 3)]:
    x = np.zeros(shape, "float32")
    w = expand_right_as(keras.ops.ones((4,)), x)             # as in compute_metrics
    err = keras.ops.mean(keras.ops.ones_like(x), axis=-1)    # as in compute_metrics
    print(shape, "->", (w * err).shape)

# (4, 6) -> (4, 4)
# (4, 8, 8, 3) -> (4, 4, 8, 8)
```

A behavioural check: reversing the weights along the batch axis (`ops.flip(w, axis=0)` inside `get_weights_for_snr`) leaves `compute_metrics(...)["loss"]` unchanged to float precision, with the noise draws replayed from the same seed state. With a per-sample reduction it changes.

## Consequences

- Training follows an unweighted noise-prediction loss multiplied by a batch-dependent scalar, `mean(w)`. The EDM weighting (`1 + exp(-log_snr) / sigma_data^2`) is effectively lost.
- The scalar is dominated by the noisiest draw in the batch. We saw training-loss outliers of 30-90x the typical value on 128x128 image targets. With gradient clipping, these outliers make the effective step size vary at random between batches.
- High-noise levels get almost no gradient relative to their intended weight. At those levels the model decides the global structure of image targets.
- At batch size 1 the result is correct, so single-sample tests do not catch it.

## Suggested fix

Reduce the error over all non-batch axes before applying the weights, and give the weights shape `(B,)`:

```python
weights_for_snr = ops.reshape(weights_for_snr, (ops.shape(x)[0],))
error = loss_mask * (noise_pred - eps_t) ** 2          # likewise for "velocity" / "F"
loss = weights_for_snr * ops.mean(error, axis=tuple(range(1, ops.ndim(error))))
loss = weighted_mean(loss, sample_weight)
```

We are using this as a subclass that overrides `compute_metrics` in the meantime. Happy to open a PR if that is useful.

## Environment

bayesflow 2.0.13, Keras 3, JAX backend (the reduction is backend-independent).

**Comment by LarsKue** [comment]
Correction to the "Consequences" section above, after checking our own training logs more carefully.

**The loss outliers I listed are not shown to come from this bug.** In our runs, data-space `DiffusionModel` training shows no loss spikes despite the bug: max/median of the training loss is about 1.2-1.4, the same as flow matching on the same data. The large outliers (18-35x the median, in about 1.5 % of steps) appear only when the diffusion model runs on an autoencoder's latent code. So they come from something specific to that setup, not from this reduction.

Why the bug is mostly benign in practice: the training times are drawn low-discrepancy, `t_i = (u0 + i / B) % 1`. So every batch covers the whole noise range, and `mean(weights)` barely changes from batch to batch. The optimized objective is then close to a constant times the unweighted noise-prediction MSE, i.e. DDPM's "simple" loss, which trains well. What is lost is the intended SNR weighting (EDM's `1 + exp(-log_snr) / sigma_data^2`, or whichever schedule is configured), not training as such.

What still holds:

- the broadcast itself;
- `loss == mean(w) * mean(err)`;
- invariance to re-pairing weights with samples;
- exact at batch size 1;
- present in every release since 2.0.4 (where `DiffusionModel` first appeared under `experimental`) through 2.0.14 and `main`, independent of the channel layout.

The effect is that any configured weighting function is silently replaced by uniform weighting in noise space. Any comparison between weighting schemes or noise schedules is therefore affected. How much this matters for sample quality presumably grows with resolution; we have not measured it yet.

**Comment by arrjon** [comment]
Thank you for raising this issue. Indeed, this is an important bug which should be fixed. The bug should only affect comparisons between different weightings. Yet, the schedule's distribution over log-SNR also acts as a weighting over noise levels, and that part was applied correctly at least.
